### PR TITLE
Add `PSR-17` Support and Remove deprecated Class/Functions

### LINKED ISSUE #25: Add PSR-17 Support
### Feature Request

|    Q        |   A
|------------ | ------
| New Feature | yes
| RFC         | yes
| BC Break    | no

#### Summary

Provide support for using [PSR-17](https://www.php-fig.org/psr/psr-17/) response factories instead of a callable for generating error and not found responses. See mezzio/mezzio#81 for details.

### PR DESCRIPTION
**Pull request by ghostwriter**
|    Q          |   A
|-------------- | ------
| New Feature   | yes

### Description

- Resolve #25

#### Added
 - [BC] PSR-17 Support

#### Changed
 - [BC] The parameter `$responseFactory` of `Laminas\Stratigility\Handler\NotFoundHandler#__construct()` changed from `callable` to a non-contravariant `Psr\Http\Message\ResponseFactoryInterface`
 - [BC] The parameter `$responseFactory` of `Laminas\Stratigility\Middleware\ErrorHandler#__construct()` changed from `callable` to a non-contravariant Psr\Http\Message\ResponseFactoryInterface
 - [BC] The parameter `$responseFactory` of `Laminas\Stratigility\Middleware\ErrorHandler#__construct()` changed from `callable` to `Psr\Http\Message\ResponseFactoryInterface`

#### Removed
 - [BC] Class `Laminas\Stratigility\Middleware\NotFoundHandler` has been deleted
 - [BC] Function `Zend\Stratigility::doublePassMiddleware` has been deleted
 - [BC] Function `Zend\Stratigility::host` has been deleted
 - [BC] Function `Zend\Stratigility::middleware` has been deleted
 - [BC] Function `Zend\Stratigility::path` has been deleted

### CODE REVIEW

**Comment by lcobucci** [comment]
The target branch should be 4.0.x, right?

File: test/Middleware/ErrorHandlerTest.php [hunk 1]
```diff
@@ -36,18 +37,17 @@ class ErrorHandlerTest extends TestCase
...
     protected function setUp(): void
     {
+        $this->request         = $this->createMock(ServerRequestInterface::class);
         $this->response        = $this->createMock(ResponseInterface::class);
-        $response              = $this->response;
-        $this->responseFactory = static fn(): ResponseInterface => $response;
-
-        $this->request        = $this->createMock(ServerRequestInterface::class);
+        $this->responseFactory = $this->createMock(ResponseFactoryInterface::class);
+        $this->responseFactory->expects(self::any())->method('createResponse')->willReturn($this->response);
```
Comment (lcobucci) [inline test/Middleware/ErrorHandlerTest.php]: We can remove the expectation and simplify a little bit, as it is set to "any":


```suggestion
        $this->responseFactory->method('createResponse')->willReturn($this->response);
```

**Review by lcobucci — COMMENTED** [review]
Tiny nitpick. Let's change the target branch and get this merged (with or without this suggestion)

**Comment by ghostwriter** [comment]
sounds good to me, @lcobucci 
I'll change the target branch once the `4.0.x` branch gets created.

File: composer.lock [hunk 2]
```diff
@@ -4,7 +4,7 @@
         "Read more about it at https://getcomposer.org/doc/01-basic-usage.md#installing-dependencies",
         "This file is @generated automatically"
     ],
-    "content-hash": "4b15d800bcbc9038df1b0d55c42b2ed9",
+    "content-hash": "af88df737b8d903a9ab666ce8afecc51",
```
Comment (lcobucci) [inline composer.lock]: Running `composer install` locally I get:

```sh
Warning: The lock file is not up to date with the latest changes in composer.json. You may be getting outdated dependencies. It is recommended that you run `composer update` or `composer update <package name>`.
```

Executing `composer update --lock` I get this diff:

```diff
-    "content-hash": "af88df737b8d903a9ab666ce8afecc51",
+    "content-hash": "2107e858ba56afe6ee5269f8e46fc834",
```

Perhaps the hash hasn't been updated after the changes in `composer.json`?

```suggestion
    "content-hash": "2107e858ba56afe6ee5269f8e46fc834",
```

File: src/Middleware/NotFoundHandler.php [hunk 3]
```diff
@@ -1,37 +0,0 @@
...
-use Laminas\Stratigility\Handler\NotFoundHandler as NotFoundRequestHandler;
-use Psr\Http\Message\ResponseInterface;
-use Psr\Http\Message\ServerRequestInterface;
-use Psr\Http\Server\MiddlewareInterface;
-use Psr\Http\Server\RequestHandlerInterface;
-
-/**
- * @deprecated Will be removed in v4 in favor of {@see \Laminas\Stratigility\Handler\NotFoundHandler}
- */
-final class NotFoundHandler implements MiddlewareInterface
```
Comment (lcobucci) [inline src/Middleware/NotFoundHandler.php:16]: Should we create a v4 dir in https://github.com/laminas/laminas-stratigility/tree/4.0.x/docs/book to have https://docs.laminas.dev/laminas-stratigility/v3/error-handlers/#handling-404-conditions updated as part of this PR?

File: src/Middleware/ErrorHandler.php [hunk 4]
```diff
@@ -70,19 +71,15 @@ class ErrorHandler implements MiddlewareInterface
...
 
     /**
-     * @param callable $responseFactory A factory capable of returning an
-     *     empty ResponseInterface instance to update and return when returning
-     *     an error response.
      * @param null|callable $responseGenerator Callback that will generate the final
      *     error response; if none is provided, ErrorResponseGenerator is used.
      */
-    public function __construct(callable $responseFactory, ?callable $responseGenerator = null)
+    public function __construct(ResponseFactoryInterface $responseFactory, ?callable $responseGenerator = null)
```
Comment (lcobucci) [inline src/Middleware/ErrorHandler.php:80]: Same thing about documentation updates (https://docs.laminas.dev/laminas-stratigility/v3/error-handlers/#handling-php-errors-and-exceptions)

**Review by lcobucci — APPROVED** [review]
I couldn't catch anything else now :+1: 
Thanks for working on this.

I'm approving because I think it's completely fine to handle the docs in another patch and wanted to let you decide what you prefer to do in this context.

**Comment by lcobucci** [comment]
@ghostwriter apologies for the huge delay. Docs look good to me, let's :ship:

**Pull request merged**